File: test_rag_playground_medical.py
#!/usr/bin/env python3
"""
RAG Studio Test Suite for asgard-medical Tenant

Tests the RAG Playground functionality on the medical tenant:
- Health checks
- Tenant verification
- Document ingestion
- Vector search
- Hybrid search
- Query responses

Usage:
    python test_rag_playground_medical.py
    MIMIR_URL=http://localhost:3002 python test_rag_playground_medical.py
"""

import os
import sys
import json
import requests
from datetime import datetime
from typing import Dict, Any, Optional
from urllib.parse import urljoin

# Configuration
BASE_URL = os.environ.get('MIMIR_URL', 'http://localhost:3002')
TENANT_ID = 'asgard-medical'
TEST_TIMEOUT = 30

# ANSI Color codes
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[32m'
    RED = '\033[31m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'

class RAGTestSuite:
    def __init__(self):
        self.tests_passed = 0
        self.tests_failed = 0
        self.session = requests.Session()

    def log(self, message: str, level: str = 'info') -> None:
        """Log message with timestamp and color."""
        timestamp = datetime.now().isoformat()

        color_map = {
            'pass': Colors.GREEN,
            'fail': Colors.RED,
            'warn': Colors.YELLOW,
            'section': Colors.CYAN,
            'info': Colors.BLUE
        }

        color = color_map.get(level, Colors.BLUE)
        print(f"{color}[{timestamp}]{Colors.RESET} {message}")

    def make_request(
        self,
        method: str,
        path: str,
        data: Optional[Dict] = None,
        expected_status: int = 200
    ) -> tuple[Optional[Dict], bool]:
        """Make HTTP request and return (response_data, success)."""
        try:
            url = urljoin(BASE_URL, path)

            if method == 'GET':
                resp = self.session.get(url, timeout=TEST_TIMEOUT)
            elif method == 'POST':
                resp = self.session.post(
                    url,
                    json=data,
                    headers={'Content-Type': 'application/json'},
                    timeout=TEST_TIMEOUT
                )
            elif method == 'DELETE':
                resp = self.session.delete(url, timeout=TEST_TIMEOUT)
            else:
                return None, False

            success = resp.status_code == expected_status

            try:
                response_data = resp.json()
            except:
                response_data = {'raw': resp.text}

            return response_data, success
        except Exception as e:
            self.log(f"Request error: {str(e)}", 'fail')
            return None, False

    def test_health_check(self) -> None:
        """Test: Service health check."""
        data, success = self.make_request('GET', '/health', expected_status=200)

        if success and data and data.get('status') == 'ok':
            self.log('✓ Health check passed', 'pass')
            self.tests_passed += 1
        else:
            self.log(f"✗ Health check failed", 'fail')
            self.tests_failed += 1

    def test_tenant_exists(self) -> None:
        """Test: Verify asgard-medical tenant exists."""
        path = f'/api/v1/tenants/{TENANT_ID}'
        data, success = self.make_request('GET', path, expected_status=200)

        if success and data:
            domain = data.get('domain', 'unknown')
            self.log(
                f"✓ Tenant '{TENANT_ID}' exists (Domain: {domain})",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Tenant '{TENANT_ID}' not found", 'fail')
            self.tests_failed += 1

    def test_list_documents(self) -> None:
        """Test: List all ingested documents."""
        path = f'/api/v1/tenants/{TENANT_ID}/ingest/documents'
        data, success = self.make_request('GET', path, expected_status=200)

        if success and isinstance(data, list):
            self.log(
                f"✓ Documents listed: {len(data)} documents found",
                'pass'
            )
            # Show first 3 documents
            for doc in data[:3]:
                doc_title = doc.get('title', 'Untitled')
                doc_id = doc.get('id', 'N/A')
                print(f"  - \"{doc_title}\" (ID: {doc_id})")
            self.tests_passed += 1
        else:
            self.log(f"✗ Failed to list documents", 'fail')
            self.tests_failed += 1

    def test_ingest_document(self) -> None:
        """Test: Ingest a test medical document."""
        test_doc = {
            "title": "RAG Playground Test Document",
            "content": """# Medical Knowledge Test Document

## Symptoms and Signs
- Fever: elevated body temperature above 38°C
- Cough: persistent, dry or productive
- Shortness of breath: dyspnea, difficulty breathing

## Common Diagnoses
- Pneumonia: infection of the lung alveoli
- Bronchitis: inflammation of the bronchial tubes
- COVID-19: coronavirus disease 2019

## Recommended Tests
- Chest X-ray: for respiratory imaging
- Blood culture: for bacterial infection detection
- PCR test: for viral pathogen identification

## Treatment Guidelines
- Supportive care: rest, hydration, monitoring
- Antibiotics: if bacterial infection confirmed
- Antiviral agents: for viral infections

## References
Based on standard medical protocols and evidence-based medicine.""",
            "source": "rag-playground-test"
        }

        path = f'/api/v1/tenants/{TENANT_ID}/ingest'
        data, success = self.make_request('POST', path, data=test_doc, expected_status=200)

        if success and data:
            doc_id = data.get('document_id', 'N/A')
            status = data.get('status', 'unknown')
            self.log(
                f"✓ Document ingested (ID: {doc_id}, Status: {status})",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Document ingestion failed", 'fail')
            self.tests_failed += 1

    def test_vector_search(self) -> None:
        """Test: Vector search query."""
        query = {
            "question": "What are the symptoms of respiratory infections?",
            "mode": "vector"
        }

        path = f'/api/v1/tenants/{TENANT_ID}/query'
        data, success = self.make_request('POST', path, data=query, expected_status=200)

        if success and data:
            mode_used = data.get('mode_used', 'unknown')
            sources_count = len(data.get('sources', []))
            answer = data.get('answer', '')

            self.log(
                f"✓ Vector search successful (Mode: {mode_used}, Sources: {sources_count})",
                'pass'
            )

            if answer:
                preview = answer[:120] + "..." if len(answer) > 120 else answer
                print(f"\n  Answer Preview:\n  \"{preview}\"\n")

            self.tests_passed += 1
        else:
            self.log(f"✗ Vector search failed", 'fail')
            self.tests_failed += 1

    def test_hybrid_search(self) -> None:
        """Test: Hybrid search query."""
        query = {
            "question": "What treatment options are available for respiratory infections?",
            "mode": "hybrid"
        }

        path = f'/api/v1/tenants/{TENANT_ID}/query'
        data, success = self.make_request('POST', path, data=query, expected_status=200)

        if success and data:
            mode_used = data.get('mode_used', 'unknown')
            sources_count = len(data.get('sources', []))

            self.log(
                f"✓ Hybrid search successful (Mode: {mode_used}, Sources: {sources_count})",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ Hybrid search failed", 'fail')
            self.tests_failed += 1

    def test_general_query(self) -> None:
        """Test: General query without mode specification."""
        query = {
            "question": "General medical information about common diseases"
        }

        path = f'/api/v1/tenants/{TENANT_ID}/query'
        data, success = self.make_request('POST', path, data=query, expected_status=200)

        if success and data:
            sources_count = len(data.get('sources', []))

            self.log(
                f"✓ General query successful (Sources: {sources_count})",
                'pass'
            )
            self.tests_passed += 1
        else:
            self.log(f"✗ General query failed", 'fail')
            self.tests_failed += 1

    def run_all_tests(self) -> None:
        """Run complete test suite."""
        self.log('=== RAG Studio Test Suite for asgard-medical ===', 'section')
        self.log(f'Base URL: {BASE_URL}', 'info')
        self.log(f'Tenant: {TENANT_ID}\n', 'info')

        # Phase 1: Infrastructure
        self.log('Phase 1: Infrastructure & Connectivity', 'section')
        self.test_health_check()

        # Phase 2: Tenant Verification
        self.log('\nPhase 2: Tenant Verification', 'section')
        self.test_tenant_exists()

        # Phase 3: Document Management
        self.log('\nPhase 3: Document Management', 'section')
        self.test_list_documents()
        self.test_ingest_document()

        # Phase 4: RAG Queries
        self.log('\nPhase 4: RAG Queries', 'section')
        self.test_vector_search()
        self.test_hybrid_search()
        self.test_general_query()

        # Summary
        self.log('\n=== Test Summary ===', 'section')
        total = self.tests_passed + self.tests_failed
        percentage = (self.tests_passed / total * 100) if total > 0 else 0

        summary_color = (
            Colors.GREEN if self.tests_failed == 0 else
            Colors.YELLOW if self.tests_failed <= 2 else
            Colors.RED
        )

        print(f"""
{summary_color}
Passed:  {self.tests_passed}/{total}
Failed:  {self.tests_failed}/{total}
Success: {percentage:.1f}%
{Colors.RESET}""")

        return 0 if self.tests_failed == 0 else 1

def main():
    """Main entry point."""
    try:
        suite = RAGTestSuite()
        exit_code = suite.run_all_tests()
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print("\n\nTest run interrupted")
        sys.exit(1)
    except Exception as e:
        print(f"Fatal error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()

File: test_rag_playground_medical.js
#!/usr/bin/env node
/**
 * RAG Studio Test Suite for asgard-medical Tenant
 *
 * Tests the RAG Playground functionality on the medical tenant:
 * - Health checks
 * - Tenant verification
 * - Document ingestion
 * - Vector search
 * - Hybrid search
 * - Query responses
 *
 * Usage:
 *   node test_rag_playground_medical.js
 *   MIMIR_URL=http://localhost:3002 node test_rag_playground_medical.js
 */

const BASE_URL = process.env.MIMIR_URL || 'http://localhost:3002';
const TENANT_ID = 'asgard-medical';
const TEST_TIMEOUT = 30000;

let testsPassed = 0;
let testsFailed = 0;

// Color codes for terminal output
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m'
};

async function log(message, type = 'info') {
  const timestamp = new Date().toISOString();
  const color = type === 'pass' ? colors.green :
                type === 'fail' ? colors.red :
                type === 'warn' ? colors.yellow :
                type === 'section' ? colors.cyan : colors.blue;
  console.log(`${color}[${timestamp}]${colors.reset} ${message}`);
}

async function testHealthCheck() {
  try {
    const res = await fetch(`${BASE_URL}/health`, { timeout: TEST_TIMEOUT });
    const data = await res.json();

    if (res.status === 200 && data.status === 'ok') {
      await log('✓ Health check passed', 'pass');
      testsPassed++;
    } else {
      await log(`✗ Health check failed: ${res.status}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Health check error: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testTenantExists() {
  try {
    const res = await fetch(`${BASE_URL}/api/v1/tenants/${TENANT_ID}`, {
      timeout: TEST_TIMEOUT
    });

    if (res.status === 200) {
      const data = await res.json();
      await log(`✓ Tenant '${TENANT_ID}' exists (Domain: ${data.domain})`, 'pass');
      testsPassed++;
    } else if (res.status === 404) {
      await log(`✗ Tenant '${TENANT_ID}' not found (404)`, 'fail');
      testsFailed++;
    } else {
      await log(`✗ Unexpected status ${res.status} when checking tenant`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error checking tenant: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testListDocuments() {
  try {
    const res = await fetch(
      `${BASE_URL}/api/v1/tenants/${TENANT_ID}/ingest/documents`,
      { timeout: TEST_TIMEOUT }
    );

    if (res.status === 200) {
      const documents = await res.json();
      await log(`✓ Documents listed: ${documents.length} documents found`, 'pass');
      if (documents.length > 0) {
        documents.slice(0, 3).forEach(doc => {
          console.log(`  - "${doc.title}" (ID: ${doc.id})`);
        });
      }
      testsPassed++;
    } else {
      await log(`✗ Failed to list documents: ${res.status}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error listing documents: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testIngestDocument() {
  try {
    const testDoc = {
      title: "RAG Playground Test Document",
      content: `# Medical Knowledge Test Document

## Symptoms and Signs
- Fever: elevated body temperature above 38°C
- Cough: persistent, dry or productive
- Shortness of breath: dyspnea, difficulty breathing

## Common Diagnoses
- Pneumonia: infection of the lung alveoli
- Bronchitis: inflammation of the bronchial tubes
- COVID-19: coronavirus disease 2019

## Recommended Tests
- Chest X-ray: for respiratory imaging
- Blood culture: for bacterial infection detection
- PCR test: for viral pathogen identification

## Treatment Guidelines
- Supportive care: rest, hydration, monitoring
- Antibiotics: if bacterial infection confirmed
- Antiviral agents: for viral infections

## References
Based on standard medical protocols and evidence-based medicine.`,
      source: "rag-playground-test"
    };

    const res = await fetch(
      `${BASE_URL}/api/v1/tenants/${TENANT_ID}/ingest`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(testDoc),
        timeout: TEST_TIMEOUT
      }
    );

    if (res.status === 200) {
      const result = await res.json();
      await log(
        `✓ Document ingested (ID: ${result.document_id}, Status: ${result.status})`,
        'pass'
      );
      testsPassed++;
    } else {
      const errorText = await res.text();
      await log(`✗ Document ingestion failed: ${res.status} - ${errorText}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error ingesting document: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testVectorSearch() {
  try {
    const query = {
      question: "What are the symptoms of respiratory infections?",
      mode: "vector"
    };

    const res = await fetch(
      `${BASE_URL}/api/v1/tenants/${TENANT_ID}/query`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(query),
        timeout: TEST_TIMEOUT
      }
    );

    if (res.status === 200) {
      const result = await res.json();
      await log(
        `✓ Vector search successful (Mode: ${result.mode_used}, Sources: ${result.sources?.length || 0})`,
        'pass'
      );

      if (result.answer) {
        console.log(`\n  Answer Preview:\n  "${result.answer.substring(0, 120)}..."\n`);
      }
      testsPassed++;
    } else {
      const errorText = await res.text();
      await log(`✗ Vector search failed: ${res.status}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error in vector search: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testHybridSearch() {
  try {
    const query = {
      question: "What treatment options are available for respiratory infections?",
      mode: "hybrid"
    };

    const res = await fetch(
      `${BASE_URL}/api/v1/tenants/${TENANT_ID}/query`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(query),
        timeout: TEST_TIMEOUT
      }
    );

    if (res.status === 200) {
      const result = await res.json();
      await log(
        `✓ Hybrid search successful (Mode: ${result.mode_used}, Sources: ${result.sources?.length || 0})`,
        'pass'
      );
      testsPassed++;
    } else {
      const errorText = await res.text();
      await log(`✗ Hybrid search failed: ${res.status}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error in hybrid search: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function testEmptyQuery() {
  try {
    const query = {
      question: "General medical information about common diseases"
    };

    const res = await fetch(
      `${BASE_URL}/api/v1/tenants/${TENANT_ID}/query`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(query),
        timeout: TEST_TIMEOUT
      }
    );

    if (res.status === 200) {
      const result = await res.json();
      await log(
        `✓ General query successful (Sources: ${result.sources?.length || 0})`,
        'pass'
      );
      testsPassed++;
    } else {
      await log(`✗ General query failed: ${res.status}`, 'fail');
      testsFailed++;
    }
  } catch (error) {
    await log(`✗ Error in general query: ${error.message}`, 'fail');
    testsFailed++;
  }
}

async function runAllTests() {
  await log('=== RAG Studio Test Suite for asgard-medical ===', 'section');
  await log(`Base URL: ${BASE_URL}`, 'info');
  await log(`Tenant: ${TENANT_ID}\n`, 'info');

  // Phase 1: Infrastructure
  await log('Phase 1: Infrastructure & Connectivity', 'section');
  await testHealthCheck();

  // Phase 2: Tenant Verification
  await log('\nPhase 2: Tenant Verification', 'section');
  await testTenantExists();

  // Phase 3: Document Management
  await log('\nPhase 3: Document Management', 'section');
  await testListDocuments();
  await testIngestDocument();

  // Phase 4: RAG Queries
  await log('\nPhase 4: RAG Queries', 'section');
  await testVectorSearch();
  await testHybridSearch();
  await testEmptyQuery();

  // Summary
  await log('\n=== Test Summary ===', 'section');
  const total = testsPassed + testsFailed;
  const percentage = total > 0 ? Math.round((testsPassed / total) * 100) : 0;

  const summaryColor = testsFailed === 0 ? colors.green :
                       testsFailed <= 2 ? colors.yellow : colors.red;

  console.log(`${summaryColor}
Passed:  ${testsPassed}/${total}
Failed:  ${testsFailed}/${total}
Success: ${percentage}%
${colors.reset}`);

  process.exit(testsFailed > 0 ? 1 : 0);
}

// Handle graceful shutdown
process.on('SIGINT', async () => {
  await log('\n\nTest run interrupted', 'warn');
  process.exit(1);
});

// Run tests
runAllTests().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});
